### PR TITLE
Add support for validation rules

### LINKED ISSUE #1342: How to disable introspection via validation rules?
**Is your feature request related to a problem? Please describe.**
I need to disable introspection for graphql, how would I do that using the method described [here](https://docs.graphene-python.org/en/latest/execution/queryvalidation/#disable-introspection)

**Describe the solution you'd like**
An example of accomplishing this would be nice

**Describe alternatives you've considered**
I am thinking to either create a subclass of `GraphQLView` or `graphene.Schema` and add the query validation there, but I don't know if that's the recommended approach

**Additional context**

### LINKED ISSUE #1472: Query Limiting Depth
**Is your feature request related to a problem? Please describe.**
Graphene has query validation about Depth limit Validator (https://docs.graphene-python.org/en/latest/execution/queryvalidation/#depth-limit-validator). 
```
validation_errors = validate(
    schema=schema.graphql_schema,
    document_ast=parse('THE QUERY'),
    rules=(
        depth_limit_validator(
            max_depth=20
        ),
    )
)
```

Would be great we will have it (depth) in django project.

**Describe the solution you'd like**
Limit depth of query. OWASP suggest add limit for security reason https://cheatsheetseries.owasp.org/cheatsheets/GraphQL_Cheat_Sheet.html#query-limiting-depth-amount

### PR DESCRIPTION
**Pull request by kiendang**
Close #1472
Close #1342

### CODE REVIEW

**Review by sjdemartini — COMMENTED** [review]
This approach LGTM! Maybe worth a quick mention in some docs/ file about how to specify `validation_rules`? And a quick test seems worthwhile before merging.

**Comment by kiendang** [comment]
@firaskafri should we release another patch version (v3.1.6) before merging this? I think this deserves a minor bump to v3.2.0 but there were a couple of small bug fixing PRs merged since v3.1.5 so I think it's a good idea to have another patch version for v3.1.

File: graphene_django/tests/test_views.py [hunk 1]
```diff
@@ -827,3 +827,69 @@ def test_query_errors_atomic_request(set_rollback_mock, client):
...
+        "data": {"__schema": {"queryType": {"name": "QueryRoot"}}}
+    }
+
+
+@pytest.mark.parametrize("url", validation_urls)
+@pytest.mark.urls("graphene_django.tests.urls_validation")
+def test_validation_disallow_introspection(client, url):
+    response = client.post(url_string(url, query="{__schema {queryType {name}}}"))
+
+    assert response.status_code == 400
```
Comment (sjdemartini) [inline graphene_django/tests/test_views.py:870]: How about we also check that the json representation of the `response` does not contain the `__schema` in its data (or data is null/empty, however it appears in practice)? That way we ensure it's not merely returning a 400 and the error message, but is also omitting the introspection data from the response as intended. (Same goes for the other 400 error tests below)

**Review by sjdemartini — APPROVED** [review]
💯 

Awesome stuff! Really appreciate the clear docs examples and tests!

Agreed with your suggestion that we should release another patch version (v3.1.6) before merging this and releasing this as a minor bump to 3.2.0

File: graphene_django/tests/test_views.py (hunk 1 above)
Comment (kiendang) [reply graphene_django/tests/test_views.py:870]: Good point!

**Review by pmcarlos — APPROVED** [review]
This looks good to me, testing it locally with another approach I used to have

**Comment by kiendang** [comment]
> This looks good to me, testing it locally with another approach I used to have

@pmcarlos Would you be able to share your current method? This PR is to be merged eventually but there's someone in Graphene Discord asking if there's another way to disable introspection.

**Comment by lee-pai-long** [comment]
Hi guys, any idea when this PR will be merged ?

**Comment by kiendang** [comment]
> Hi guys, any idea when this PR will be merged ?

This would be part of the `3.2.0` release. The plan is to release another patch version `3.1.6` with the current `main` branch, then we'll merge this PR and do a `3.2.0` later. @firaskafri is there any road block on making a `3.1.6` release?

**Comment by firaskafri** [comment]
> > Hi guys, any idea when this PR will be merged ?
> 
> This would be part of the `3.2.0` release. The plan is to release another patch version `3.1.6` with the current `main` branch, then we'll merge this PR and do a `3.2.0` later. @firaskafri is there any roadblock on making a `3.1.6` release?

Hello @kiendang not at all. 
I just released 3.1.6 and merging this to main now.

**Review by firaskafri — APPROVED** [review]

**Comment by kiendang** [comment]
> > > Hi guys, any idea when this PR will be merged ?
> > 
> > 
> > This would be part of the `3.2.0` release. The plan is to release another patch version `3.1.6` with the current `main` branch, then we'll merge this PR and do a `3.2.0` later. @firaskafri is there any roadblock on making a `3.1.6` release?
> 
> Hello @kiendang not at all. I just released 3.1.6 and merging this to main now.

@firaskafri  Awesome. Thanks! We're good to go for `3.2.0` as well.

**Comment by lee-pai-long** [comment]
Hi @kiendang any example on how to implement both rules ? I tried to follow the graphene documentation:
```
from graphene import Schema
from .auto import schema_operations_builder
from graphql import validate, parse
from graphene.validation import depth_limit_validator, DisableIntrospection

schema = Schema(query=ALL_QUERIES,mutation=ALL_MUTATIONS)

validation_errors = validate(
    schema=schema.graphql_schema,
    document_ast=parse('THE QUERY'),
    rules=(
        depth_limit_validator(max_depth=10),
        DisableIntrospection
    )
)
```

But I get an error `ModuleNotFoundError: No module named 'graphene.validation'`

**Comment by kiendang** [comment]
@lee-pai-long hmm that's weird. It comes with Graphene. We have a test for it in this repo and it works fine. Alternatively the introspection disabling rule is also implemented in `graphql-core`. You can do

```py
from graphql.validation.rules.custom.no_schema_introspection import NoSchemaIntrospectionCustomRule
```

`NoSchemaIntrospectionCustomRule` is the same as `graphene.validation.DisableIntrospection`. However `depth_limit_validator` is only available in Graphene though.

**Comment by lee-pai-long** [comment]
Hi @kiendang thanks for you help, actually my error was that I didn't have the right version of `graphene`, after upgrading it and the other dependencies I can import the `validation` module.

**Pull request merged**